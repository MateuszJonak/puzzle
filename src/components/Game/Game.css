.game {
  margin: auto;
  max-width: 600px;
}

.tiles-box-container {
  width: 100%;
  height: 200px;
}

.actions-container {
  padding-bottom: 10px;
}

.board-container{
  padding-bottom: 10px;
}

.btn {
  overflow: hidden;

  margin-bottom: 5;
  padding: 12px 12px;

  cursor: pointer;
  user-select: none;
  transition: all 150ms linear;
  text-align: center;
  white-space: nowrap;
  text-decoration: none !important;
  text-transform: none;
  text-transform: capitalize;

  color: #fff;
  border: 0 none;
  border-radius: 4px;

  font-size: 13px;
  font-weight: 500;
  line-height: 1.3;

  -webkit-appearance: none;
  -moz-appearance:    none;
  appearance:         none;
  
  box-shadow: 2px 5px 10px var(--color-smoke);
}

.btn:hover {
  transition: all 150ms linear;

  opacity: .85;
}

.btn:active {
  transition: all 150ms linear;
  opacity: .75;
}

.btn:focus {
  outline: 1px dotted #959595;
  outline-offset: -4px;
}

.btn-regular {
  color: #202129;
  background-color: #f2f2f2;
}

.btn-regular:hover {
  color: #202129;
  background-color: #e1e2e2;
  opacity: 1;
}

.btn-regular:active {
  background-color: #d5d6d6;
  opacity: 1;
}