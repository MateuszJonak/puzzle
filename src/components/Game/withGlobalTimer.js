import React, { Component } from 'react';
import { connect } from 'react-redux';
import actions from '../../store/game/actions';

const withGlobalTimer = WrappedComponent => {
  class GlobalTimer extends Component {
    globalTimerFinish = () => {
      this.stopTimer();
      this.handleTick();
    };

    startTimer = () => {
      this.timer = setInterval(this.handleTick, 1000);
    };

    stopTimer = () => {
      clearInterval(this.timer);
    };

    handleTick = () => {
      this.props.elapsedUpdate({ elapsed: this.props.localElapsed });
    };

    render() {
      const { elapsedUpdate, ...restProps } = this.props;
      return (
        <WrappedComponent
          globalTimerStart={this.startTimer}
          globalTimerStop={this.stopTimer}
          globalTimerFinish={this.globalTimerFinish}
          {...restProps}
        />
      );
    }
  }
  const mapPropsToState = () => ({});
  const mapDispatchToProps = {
    elapsedUpdate: actions.elapsed.update,
  };

  return connect(
    mapPropsToState,
    mapDispatchToProps,
  )(GlobalTimer);
};

export default withGlobalTimer;
