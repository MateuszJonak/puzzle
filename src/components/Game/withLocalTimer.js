import React, { Component } from 'react';

const withLocalTimer = WrappedComponent => {
  class LocalTimer extends Component {
    constructor(props) {
      super(props);

      this.state = { startTime: null, localElapsed: 0, lastElapsed: 0 };
    }

    localTimerStart = ({ lastElapsed = 0 } = {}) => {
      this.setState({ lastElapsed, startTime: new Date() });
      this.startTimer();
    };

    localTimerFinish = () => {
      this.setState({ startDate: null, localElapsed: 0, lastElapsed: 0 });
      this.stopTimer();
    };

    startTimer = () => {
      this.timer = setInterval(this.handleTick, 50);
    };

    stopTimer = () => {
      clearInterval(this.timer);
    };

    handleTick = () => {
      const { lastElapsed, startTime } = this.state;
      this.setState({
        localElapsed: lastElapsed + (new Date() - startTime),
      });
    };

    render() {
      return (
        <WrappedComponent
          localTimerStart={this.localTimerStart}
          localTimerStop={this.stopTimer}
          localTimerFinish={this.localTimerFinish}
          {...this.state}
          {...this.props}
        />
      );
    }
  }

  return LocalTimer;
};

export default withLocalTimer;
