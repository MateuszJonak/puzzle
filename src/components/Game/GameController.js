import React, { Component } from 'react';
import { connect } from 'react-redux';
import { compose } from 'redux';
import PropTypes from 'prop-types';
import moment from 'moment';
import momentDurationFormatSetup from 'moment-duration-format';
import actions from '../../store/game/actions';
import {
  getIsRunning,
  getIsFinished,
  getElapsed,
} from '../../store/game/selectors';
import withLocalTimer from './withLocalTimer';
import withGlobalTimer from './withGlobalTimer';

momentDurationFormatSetup(moment);

export class GameController extends Component {
  componentDidMount() {
    const { isRunning, elapsed } = this.props;
    if (isRunning) {
      this.timersStart({ lastElapsed: elapsed });
    }
  }

  componentWillUnmount() {
    this.timersStop();
  }

  gameStart = () => {
    const { actionGameStart, isRunning, isFinished } = this.props;
    if (!isFinished && !isRunning) {
      this.timersStart();
      actionGameStart();
    }
  };

  gameReset = () => {
    const { actionGameReset } = this.props;
    this.timersReset();
    actionGameReset();
  };

  gameFinish = () => {
    const { actionGameFinish } = this.props;
    this.timersFinish();
    actionGameFinish();
  };

  timersStart({ lastElapsed } = {}) {
    const { localTimerStart, globalTimerStart } = this.props;
    localTimerStart({ lastElapsed });
    globalTimerStart();
  }

  timersStop() {
    const { localTimerStop, globalTimerStop } = this.props;
    localTimerStop();
    globalTimerStop();
  }

  timersReset() {
    const { localTimerFinish, globalTimerStop } = this.props;
    localTimerFinish();
    globalTimerStop();
  }

  timersFinish() {
    const { globalTimerFinish, localTimerFinish } = this.props;
    globalTimerFinish();
    localTimerFinish();
  }

  render() {
    const { localElapsed, isFinished, elapsed, children } = this.props;

    const duration = isFinished
      ? moment.duration(elapsed).format('mm:ss:SS')
      : moment.duration(localElapsed).format('mm:ss:SS');

    return (
      <React.Fragment>
        {children({
          gameStart: this.gameStart,
          gameFinish: this.gameFinish,
          gameReset: this.gameReset,
          duration,
        })}
      </React.Fragment>
    );
  }
}
const mapPropsToState = state => ({
  isRunning: getIsRunning(state),
  isFinished: getIsFinished(state),
  elapsed: getElapsed(state),
});
const mapDispatchToProps = {
  actionGameStart: actions.game.start,
  actionGameFinish: actions.game.finish,
  actionGameReset: actions.game.reset,
};

const propsFunctions = [
  'actionGameStart',
  'actionGameFinish',
  'actionGameReset',
  'localTimerStart',
  'localTimerStop',
  'localTimerFinish',
  'globalTimerStart',
  'globalTimerStop',
  'globalTimerFinish',
];

const propsFunctionsPropTypes = propsFunctions.reduce(
  (result, propType) => ({ ...result, [propType]: PropTypes.func }),
  {},
);

const propsFunctionsDefaultProps = propsFunctions.reduce(
  (result, propType) => ({ ...result, [propType]: () => {} }),
  {},
);

GameController.propTypes = {
  isRunning: PropTypes.bool,
  isFinished: PropTypes.bool,
  elapsed: PropTypes.number,
  localElapsed: PropTypes.number,
  children: PropTypes.func,
  ...propsFunctionsPropTypes,
};

GameController.defaultProps = {
  isRunning: false,
  isFinished: false,
  elapsed: 0,
  localElapsed: 0,
  children: () => {},
  ...propsFunctionsDefaultProps,
};

export default compose(
  withLocalTimer,
  withGlobalTimer,
  connect(
    mapPropsToState,
    mapDispatchToProps,
  ),
)(GameController);
