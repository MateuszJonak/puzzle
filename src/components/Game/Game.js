import React, { Component } from 'react';
import { Board } from '../Board';
import { TilesBox } from '../TilesBox';
import GameController from './GameController';
import './Game.css';

class Game extends Component {
  render() {
    return (
      <GameController>
        {({ gameStart, gameFinish, gameReset, duration }) => (
          <div className="game">
            <div className="game-timer">
              <h2>{duration}</h2>
            </div>
            <div className="actions-container">
              <button className="btn btn-regular" onClick={() => gameReset()}>
                restart
              </button>
            </div>
            <div className="board-container">
              <Board onFillEnd={gameFinish} />
            </div>
            <div className="tiles-box-container">
              <TilesBox onBeginFirstDrag={gameStart} />
            </div>
          </div>
        )}
      </GameController>
    );
  }
}

export default Game;
