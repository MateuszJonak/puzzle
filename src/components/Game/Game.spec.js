import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Game from './Game';
const mockDuration = 600;
// const mockGameController = ({ children }) =>
// children({ duration: mockDuration });
jest.mock('./GameController', () => ({ children }) =>
  children({ duration: mockDuration }),
);

describe('Game', () => {
  let tree;
  const shallowEnzyme = (props = {}) => shallow(<Game {...props} />);

  beforeEach(() => {
    tree = shallowEnzyme();
  });

  it('should have instance of GameController', () => {
    const instance = tree.instance();

    expect(instance).toBeInstanceOf(Game);
  });

  it('should render component correctly with props', () => {
    expect(toJson(tree.dive())).toMatchSnapshot();
  });
});
