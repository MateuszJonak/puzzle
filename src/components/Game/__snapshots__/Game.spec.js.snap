// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game should render component correctly with props 1`] = `
<div
  className="game"
>
  <div
    className="game-timer"
  >
    <h2>
      600
    </h2>
  </div>
  <div
    className="actions-container"
  >
    <button
      className="btn btn-regular"
      onClick={[Function]}
    >
      restart
    </button>
  </div>
  <div
    className="board-container"
  >
    <Connect(BoardContainer) />
  </div>
  <div
    className="tiles-box-container"
  >
    <Connect(TilesBoxContainer) />
  </div>
</div>
`;
